src/array_api_stubs/_draft/creation_functions.py

```diff
@@ -214,19 +215,38 @@ def eye(
     """
 
 
-def from_dlpack(x: object, /) -> array:
+def from_dlpack(
+    x: object, /, *,
+    device: Optional[device] = None,
+    copy: Optional[bool] = None
+) -> Union[array, Any]:
     """
     Returns a new array containing the data from another (array) object with a ``__dlpack__`` method.
 
     Parameters
     ----------
     x: object
         input (array) object.
+    device: Optional[device]
+        device on which to place the created array. If ``device`` is ``None`` and ``x`` supports DLPack, the output array device must be inferred from ``x``. Default: ``None``.
+
+        The v2023.12 standard only mandates that a compliant library must offer a way for ``from_dlpack`` to create an array on CPU (using
+        the library-chosen way to represent the CPU device - ``kDLCPU`` in DLPack - e.g. a ``"CPU"`` string or a ``Device("CPU")`` object).
+        If the compliant library does not support the CPU device and needs to outsource to another (compliant) array library, it may do so
+        with a clear user documentation and/or run-time warning. If a copy must be made to enable this, and ``copy`` is set to ``False``,
+        the function must raise ``ValueError``.
+
+        Other kinds of devices will be considered for standardization in a future version.
+    copy: Optional[bool]
+        boolean indicating whether or not to copy the input. If ``True``, the function must always copy. If ``False``, the function must never copy and must raise a ``BufferError`` in case a copy would be necessary (e.g. the producer disallows views). If ``None``, the function must reuse existing memory buffer if possible and copy otherwise. Default: ``None``.
+
+        If a copy is needed, the stream over which the copy is performed must be taken from the consumer, following the DLPack protocol (see :meth:`array.__dlpack__`).
 
     Returns
     -------
-    out: array
-        an array containing the data in `x`.
+    out: Union[array, Any]
+        an array containing the data in ``x``. In the case that the compliant library does not support the given ``device`` out of box
+        and must oursource to another (compliant) library, the output will be that library's compliant array object.
```

Comment:
Thinking out loud: I would like to highlight this treatment. When I wrote this, I had this use case in mind:
```python
xp = a. __array_namespace__()
# to avoid unnecessary distraction, I don't use a "cpu" string here...
b = xp.from_dlpack(a, copy=True, device=<some_object_representing_cpu>)
```
because the original issue asked if a portable `.to_device()` method is possible. This is doable, at the expense of 
1. crossing two array namespaces (since `b` may not be of the same type `xp.array` as `a`, ex: CuPy -> NumPy) 
2. the extra burden for the namespace `xp` to maintain an device object representing `kDLCPU`, which may or may not be feasible. 

Do people feel better if we insist namespace leak is not allowed? So instead we would require users (and libraries!) to explicitly use up to two namespaces, e.g.
```python
# xp1 and xp2 may or may not be the same, e.g.
#   1. xp1 = cp, xp2 = np
#   2. xp1 = xp2 = torch 
a = xp1.random.random(...)
b = xp2.from_dlpack(a, copy=True, device=another_xp2_array.device)
```
which would bring us back to the explicit N-to-N library conversion hell (ex: will we have to write this code in SciPy?!), but potentially reduce confusion at run time and avoid the two costs noted above (no namespace leak, and no redundant device object).

Reply:
So for the `from_dlpack` I am not sure I am too worried about `device=` ending up specific to `xp2`, there should be a mapping from an `xp2` device to a dlpack device, but the user doesn't need to know/understand it.

The important part is that `xp2` translates it to a dlpack device (or a dlpack device tuple `(device_id, device_num)`, depending on what we specify right now).

I suspect the first usage is probably `numpy.from_dlpack(obj)` or `cupy.from_dlpack(obj)` to work.  I.e. we have a reference library to use because:
* We know speed is irrelevant and we need additional capabilities (say some other library that only supports numpy/cupy). 
* This is testing code.

In those common cases the author already knows how to hardcode what they want.  (It is a bit funny that the user has to pass `device="cpu"` to allow a cross-device copy, but it seems fair.)

---

Thinking about synchronization again, I think we have two obvious choices for what `stream` means: It can be in terms of originating device or in terms of target device.

I am starting to think that defining synchronization based on the *target* device (i.e. the one passed in) may have some advantages:

* If the target is CPU, we don't have a synchronization mechanism, so it is obvious that we synchronize.
* It is at least somewhat nice to know that `dldevice=kCUDA, stream=CudaStream` always makes sense!

That means I can't specify synchronized to-device copy if the target is `CPU`.  But, I am not sure if that matters too much for Cuda because we do have the cuda managed and cuda pinned memory device types?  And those _could_ specify to use the same semantics as the CUDA one (with the caveat of maybe defaulting to full synchronization).

Reply:
> So for the `from_dlpack` I am not sure I am too worried about `device=` ending up specific to `xp2`, there should be a mapping from an `xp2` device to a dlpack device, but the user doesn't need to know/understand it.

I agree. Neither (1) nor (2) seem to be a concern, and I don't quite understand the "which would bring us back to the explicit N-to-N library conversion hell". The `device=` keyword is already specific to a particular library everywhere, e.g. all creation functions and `to_device`. So nothing changes here regarding how device specifiers work. The only way to obtain them is with `some_array.device`, and hence the two code examples given above are already essentially the same.

I think there is nothing that can be changed or is of particular concern for this topic. So I suggest we resolve this comment, and copy the synchronization-related part of @seberg's comment to a non-inline comment (it's a different topic).

Reply:
> The only way to obtain them is with some_array.device, and hence the two code examples given above are already essentially the same.

Reading [the thread](https://github.com/data-apis/array-api/pull/741#discussion_r1485213049) below, I am not sure I agree. In order for `from_dlpack()` to allow a way to copy data to host, each library needs its own CPU identifier (which we leave unspecified, except noting in this PR that it is allowed).

> and I don't quite understand the "which would bring us back to the explicit N-to-N library conversion hell".

When I wrote this, I had in mind the downstream libraries like SciPy and scikit-learn relying on this to work. They would have to detect `xp1` namespace (which they already did), and come up with a suitable `xp2` choice that allows them to move data to CPU.

Reply:
> and come up with a suitable `xp2` choice that allows them to move data to CPU.

that's not a problem: for scipy/sklearn that is always going to be numpy. 

> each library needs its own CPU identifier (which we leave unspecified

Sorry, I meant "obtain portably". You can use `device=mylib.cpu` indeed, but that's outside the standard. Using features provided by the standard, one can write `x.device`. E.g. in test cases for SciPy:
```python
def test_func(xp):  # test fixture xp, can be any library 
    x = xp.asarray(...)
    actual = func(x)
    expected = np.asarray([0.51, 3.54])
    assert_allcllose(actual, expected)   # problem due to mismatching namespaces
```
gets modified to:
```python
def new_assert_allclose(actual, expected, ...):
    actual = np.from_dlpack(actual, expected.device)
    assert_allcllose(actual, expected)
```

I'm still not sure why your second code example would be a problem. It looks the way I expect it, and should be usable enough. The first one works fine too though.

Reply:
My feeling is each of us applied some assumptions regarding portability. Let me sketch my perception.  

I left the 2nd code example (which would certainly work, I am not debating whether it would, but that it looks uglier) because for the 1st one to work it requires the highlighted escape clause (regarding cross namespaces), not because they are separate, independent considerations. 

Think of `xp` being CuPy, the 1st snippet
```python
# a is a cupy.ndarray
b = cp.from_dlpack(a, copy=True, device=cp.CPU)
# b is a numpy.ndarray
```
means `b` must be `np.ndarray`. There is no way that `b` can still be `cp.ndarray` while residing on CPU. CuPy simply disallows that.

But this is my definition of portability: When #626 was raised, it was proposed to use a method (either modifying `.to_device()` or adding `.to_numpy()`) that should always work. This PR simply says that instead of modifying a method we modify a free function. If we say the 1st snippet would not work for CuPy, it doesn't look good as far as portability is concerned, although this use case does mean that a non-portable identifier `cupy.CPU` has to be defined and used.

The other definition of portability emerges as "the `device=` argument must always be the `.device` attribute from an existing consumer array, but `from_dlpack` doesn't have to always work". Then, the slightly modified 1st snippet becomes
```python
# a is a cupy.ndarray
b = np.from_dlpack(a, copy=True, device=another_np_arr.device)
# b is a numpy.ndarray
```
and certainly it'd work without any non-portable device identifier. My complaint about its ugliness is that in a production, downstream code it means we have the N-to-N conversion to consider. I did not expect at all 
> that's not a problem: for scipy/sklearn that is always going to be numpy.

because I assumed what's discussed here could be applicable to actual array-api compliant implementation as well, not just a simple test helper. That "`xp2` is always numpy" is nothing portable either. I think we agreed NumPy should not enjoy a special status. (Also, if a torch GPU tensor is in use, in this example we should expect a torch cpu tensor is returned, not a numpy array.)

If it's all about test helpers that are aimed after in this PR and the original issue, we should not need any standardization, just ask each library to duplicate a few lines of code (from your above illustration). It's not a big deal, I am sure all libraries have done much worse hacks 😄 

@rgommers WDYT?

Reply:
> means `b` must be `np.ndarray`. There is no way that `b` can still be `cp.ndarray` while residing on CPU. CuPy simply disallows that.

Yes, so CuPy must raise, there is no other reasonable choice. Returning a non-native array type from a user-facing API seems like an anti-pattern that has no precedent and would be very surprising. 

> But this is my definition of portability: When #626 was raised, it was proposed to use a method (either modifying `.to_device()` or adding `.to_numpy()`) that should always work. [...]

The goal in gh-626, which came from SciPy, is "make it possible for any array to be turned into a numpy array". For this, I'd expect `np.from_dlpack(...)`. Which is your second code example in the comment right above.

> My complaint about its ugliness is that in a production, downstream code it means we have the N-to-N conversion to consider.

Why do you return to N-to-N? I think the arguments you gave don't apply here. In `np.from_dlpack(a, copy=True, device=another_np_arr.device)`, the `from_dlpack` function and the `device=...` argument must be from the same library, that is all. That is how `device` keyword work in all other functions too. And it's not numpy-specific, because:

> That "`xp2` is always numpy" is nothing portable either. I think we agreed NumPy should not enjoy a special status. (Also, if a torch GPU tensor is in use, in this example we should expect a torch cpu tensor is returned, not a numpy array.)

This doesn't sound right to me. It can equally be this:
```python
# a is a cupy.ndarray
b = torch.from_dlpack(a, copy=True, device=torch.device('cpu'))
# b is a torch.Tensor on cpu
```

To me, the point of this exercise is that:
1. the library implementer `xp` accepts its own device specifier in `from_dlpack` (for now CPU-only, but can be extended later as we discussed)
2. translates that to `kDLCPU` under the hood and passes it to the producer library
3. that producer library returns a DLPack capsule with data in host memory
4. hence, if `xp` supports CPU arrays, there is _always_ a way to get that from an arbitrary library, _even if_ that library doesn't support CPU arrays natively.

There is no N-to-N problem, because what is exchanged between libraries is only the portable `kDLCPU` enum value.

Reply:
Edited my above comment for clarity. We focused on CPU, but there is even a way to go from (for example) a CuPy array on a CUDA device to a OneAPI array on a SYCL device, _via_ host memory.

Reply:
> The goal in [gh-626](https://github.com/data-apis/array-api/issues/626), which came from SciPy, is "make it possible for any array to be turned into a numpy array". For this, I'd expect `np.from_dlpack(...)`. Which is your second code example in the comment right above.
> 

I have read #626 multiple times while preparing this PR, and my take was that "turned into a **numpy array**" was never the ask. As highlighted in that issue title and also noted by you later, the ask was to have something that's accessible from host (or, using Sasha's words, to the Python interpreter). Whether or not it's a NumPy array is not a  big deal. 

> Why do you return to N-to-N? 

I insist that N-to-N is a problem because we're forcing (or "encouraging", if this choice of wording helps) a downstream library code to decide what is their favorite `xp2` pick (`xp1` is from user input so they don't get to choose). This question is still unanswered. 

I have no objection/concern in how `xp1` and `xp2` in the 2nd snippet should interact and pan out in an actual implementation, I think implementation-wise and semantics-wise it is crystal clear so I don't think I have anything extra to say. My objection is why we're thinking about it at all. Let me ask this way using your example.
> This doesn't sound right to me. It can equally be this:
> ```python
> # a is a cupy.ndarray
> b = torch.from_dlpack(a, copy=True, device=torch.device('cpu'))
> # b is a torch.Tensor on cpu
> ```

 If `xp1` is cupy, how would a downstream library know that it can use torch as `xp2`? Either they insist on pre-selecting a default array library for `xp2`, or they come up with a bunch of if-else branches to encode this logic (try torch, then try jax, they try numpy, then ...). How is this justifying that the standard is array-library agnostic? 

If this question does not get across, I am afraid we will need a call to resolve this. I do not have a better, alternative way to phrase this question...

Reply:
> Whether or not it's a NumPy array is not a big deal.

agreed, it works the same for any library - as long as it supports exchanging via host memory.

> downstream library code to decide what is their favorite `xp2` pick (`xp1` is from user input so they don't get to choose

In practice this is almost always the case, but you can even avoid it (xref your `# xp1 and xp2 may or may not be the same, e.g.` example above).

> My objection is why we're thinking about it at all

Okay. The direct reason for gh-626 was that, as of now, it was not possible at all to work with two libraries that don't have a common device (e.g., NumPy and CuPy) without special-casing the library. After this PR, it will be. 

> If `xp1` is cupy, how would a downstream library know that it can use torch as `xp2`? 

It will always work with any library, on any device, as long as they can both interact with host memory. 

> Either they insist on pre-selecting a default array library for `xp2`, or they come up with a bunch of if-else branches to encode this logic (try torch, then try jax, they try numpy, then ...). How is this justifying that the standard is array-library agnostic?

No, no pre-selecting is needed, nor is an if-else block. Your code from your first comment just works:
```python
a = xp1.random.random(...)
b = xp2.from_dlpack(a, copy=True, device=another_xp2_array.device)
```

I'll make it even more explicit. Say you have no preferred library and want to write super-generic code - some function that takes two arrays that are from different libraries. Right now, you just can't do that. After this PR, you can:
```python
def func(x, y):
    xp = x.__array_namespace__()
    xp_other = y.__array_namespace__()

    # Other functions than `from_dlpack` only work if both arrays are from the same library. So if
    # `y` is from a different one than `x`, let's convert `y` into an array of the same type as `x`:
    if not xp == xp_other:
        # this always works; note that if `x.device` isn't CPU, it may incur two data copies
        # (from y.device to host on the producer side, then from host to x.device on consumer side).
        y = xp.from_dlpack(y, copy=True, device=x.device)

    # From now on use `xp.xxx` functions
```

> If this question does not get across, I am afraid we will need a call to resolve this

If the above still isn't addressing your point, let's do a quick call later today or tomorrow so we can talk through it?

Reply:
Thanks @rgommers. Your super-generic code snippet looks nice, I included it in the docstring. Though I think it just kick the answer to "how to pick `xp2`?" down the road. Somewhere, at some point, the choice has to be made to provide the array `y`.

Nevertheless, I think if people are happy with the PR as-is, I shouldn't be blocking for two reasons:
1. Making it strict (no namespace leak) in the beginning is better than relaxed at first and tightened later.
2. If I want to make `cupy.from_dlpack(arr_cp, copy=True, device=cupy.CPU)` work and not raise an exception (not necessary a good idea, ofc), I still can, it's just beyond what's standardized.

One minor note on the code comment:
>         # this always works; note that if `x.device` isn't CPU, it may incur two data copies
>         # (from y.device to host on the producer side, then from host to x.device on consumer side).

I think so far we've not mentioned anything about double copy, so I didn't include this comment, though it's not disallowed either. If a library wants to pick a complex route, they can.